File: LeetCode/918MaximumCircularSubArray/MaximumCircularSubArray.py
lass Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        # Kadane's algorithm -> Time:O(n), Space:O(1)
        global_max, global_min = nums[0], nums[0]
        curr_max, curr_min, total = 0, 0, 0
        for i in nums:
            curr_max = max(curr_max + i, i)
            curr_min = min(curr_min + i, i)
            total += i
            global_max = max(global_max, curr_max)
            global_min = min(global_min, curr_min)
        return max(global_max, total - global_min) if global_max >= 0 else global_max

File: LeetCode/918MaximumCircularSubArray/MaximumCircularSubArray.cpp
lass Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        // Kadane's algorithm -> Time:O(n), Space:O(1)
        int globalMax = nums[0], globalMin = nums[0], total = 0, currMax = 0, currMin = 0;
        for (int i = 0; i < nums.size(); i++) {
            currMax = max(currMax + nums[i], nums[i]);
            currMin = min(currMin + nums[i], nums[i]);
            total += nums[i];
            globalMax = max(globalMax, currMax);
            globalMin = min(globalMin, currMin);
        }
        if (globalMax < 0) {
            return globalMax;
        }
        return max(globalMax, total - globalMin);
    }
};
